File: jetbrains-plugin/src/main/kotlin/com/oneide/SyncService.kt
package com.oneide

import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.wm.WindowManager
import com.oneide.models.IdeMetaData
import com.oneide.models.Role
import com.oneide.models.State
import com.oneide.services.ClusterService
import com.oneide.services.IdeConnector
import com.oneide.services.StateService
import com.oneide.services.cluster.ActionRegistry
import com.oneide.utils.Logger

/**
 * The core service that orchestrates One-IDE synchronization for a specific project.
 *
 * Responsibilities:
 * - Initializes and manages lifecycle of StateService.
 * - Subscribes to ClusterService role changes and user activity.
 * - Holds the project metadata ([IdeMetaData]).
 * - Manages the global enabled/disabled state for synchronization.
 * - Updates status bar widgets across open projects when state changes.
 */
@Service(Service.Level.PROJECT)
class SyncService(private val project: Project) : Disposable {
    val metaData = IdeMetaData.getInstance(project)
    private val logger = Logger.withMetaData(metaData)

    private val clusterService = ClusterService.getInstance(project)
    private val stateService = StateService(metaData)
    private val ideConnector: IdeConnector = clusterService.getIdeConnector()

    private var unsubscribeRoleChange: (() -> Unit)? = null
    private var unsubscribeLeaderUserActivity: (() -> Unit)? = null
    private var lastState: State? = null

    var isEnabled: Boolean = true
        set(value) {
            field = value
            logger.info("One-IDE Sync enabled: $value")
            updateAllStatusBars()
        }

    init {
        logger.info("One-IDE SyncService initialized. SourceID: ${metaData.id}")

        // Setup state received callback
        stateService.setOnStateReceived { state -> onStateReceived(state) }

        // Subscribe to role changes
        unsubscribeRoleChange = clusterService.addRoleChangeListener { role ->
            handleRoleChange(role)
        }

        // Subscribe to user activity only when LEADER role
        // The Leader role fires this action, SyncService listens
        unsubscribeLeaderUserActivity = clusterService.addAction(Role.LEADER, ActionRegistry.ACTION_USER_ACTIVITY) {
            if (isEnabled) {
                publishCurrentState()
            }
        }

        // Handle initial role
        handleRoleChange(clusterService.getRoleType())
    }

    private fun handleRoleChange(role: Role) {
        when (role) {
            Role.LEADER -> {
                stateService.stopWatching()
                // Publish current state immediately when becoming leader
                publishCurrentState()
            }
            Role.FOLLOWER -> {
                stateService.startWatching()
            }
            Role.CANDIDATE -> {
                // No action needed for candidate
            }
        }
    }

    private fun publishCurrentState() {
        val state = ideConnector.captureState()
        if (state != lastState) {
            logger.info("State changed, publishing new state")
            stateService.publishState(state, clusterService.getNodeId())
            lastState = state
        }
    }

    private fun onStateReceived(state: State) {
        if (clusterService.getRoleType() == Role.FOLLOWER && isEnabled) {
            ideConnector.applyState(state)
        }
    }

    override fun dispose() {
        logger.info("Disposing SyncService for ${project.name}")
        unsubscribeRoleChange?.invoke()
        unsubscribeLeaderUserActivity?.invoke()
        stateService.dispose()
    }

    companion object {
        fun getInstance(project: Project): SyncService = project.getService(SyncService::class.java)
    }

    private fun updateAllStatusBars() {
        ApplicationManager.getApplication().invokeLater {
            val projects = ProjectManager.getInstance().openProjects
            for (project in projects) {
                val statusBar = WindowManager.getInstance().getStatusBar(project)
                statusBar?.updateWidget("OneIDEStatusBar")
            }
        }
    }
}

File: vscode-extension/src/syncService.ts
import * as vscode from 'vscode';
import { IdeMetaData } from './IdeMetaData';
import { ClusterService, RoleType } from './services/ClusterService';
import { StateService } from './services/StateService';
import { IdeConnector } from './services/IdeConnector';
import { State } from './types';
import { Logger } from './logger';
import { ActionRegistry } from './services/cluster/ActionRegistry';

/**
 * The core service that orchestrates One-IDE synchronization.
 *
 * Responsibilities:
 * - Subscribes to ClusterService role changes and user activity.
 * - Manages StateService watching based on role.
 * - Captures and publishes state when Leader.
 * - Applies received state when Follower.
 * - Manages the global enabled/disabled state for synchronization.
 * - Updates status bar widgets.
 */
export class SyncService implements vscode.Disposable {
    private sourceId: string;
    private isEnabled: boolean = true;
    private statusBarItem: vscode.StatusBarItem;

    private clusterService: ClusterService;
    private stateService: StateService;
    private ideConnector: IdeConnector;

    private unsubscribeRoleChange: (() => void) | null = null;
    private unsubscribeLeaderUserActivity: (() => void) | null = null;
    private lastStateJson: string | null = null;

    constructor(
        clusterService: ClusterService,
        stateService: StateService,
        ideConnector: IdeConnector
    ) {
        const meta = IdeMetaData.getInstance();
        this.sourceId = meta.id;
        Logger.setMetaData(meta);
        Logger.log(`Source ID: ${this.sourceId}`);
        
        this.clusterService = clusterService;
        this.stateService = stateService;
        this.ideConnector = ideConnector;

        this.statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 100);
        this.statusBarItem.command = 'one-ide.toggle';
        this.updateStatusBar();
        this.statusBarItem.show();

        // Setup state received callback
        this.stateService.setOnStateReceived((state) => this.onStateReceived(state));

        // Subscribe to role changes
        this.unsubscribeRoleChange = this.clusterService.addRoleChangeListener((role) => {
            this.handleRoleChange(role);
        });

        // Subscribe to user activity only when LEADER role
        // The Leader role fires this action, SyncService listens
        this.unsubscribeLeaderUserActivity = this.clusterService.addAction(
            RoleType.LEADER,
            ActionRegistry.ACTION_USER_ACTIVITY,
            () => {
                if (this.isEnabled) {
                    this.publishCurrentState();
                }
            }
        );

        // Handle initial role
        this.handleRoleChange(this.clusterService.getRoleType());
    }

    private handleRoleChange(role: RoleType) {
        switch (role) {
            case RoleType.LEADER:
                this.stateService.stopWatching();
                // Publish current state immediately when becoming leader
                this.publishCurrentState();
                break;
            case RoleType.FOLLOWER:
                this.stateService.startWatching();
                break;
            case RoleType.CANDIDATE:
                // No action needed for candidate
                break;
        }
    }

    private async publishCurrentState() {
        const state = await this.ideConnector.captureState();
        const stateJson = JSON.stringify(state);

        if (stateJson !== this.lastStateJson) {
            Logger.log('State changed, publishing new state');
            await this.stateService.publishState(state, this.clusterService.getNodeId());
            this.lastStateJson = stateJson;
        }
    }

    private async onStateReceived(state: State) {
        if (this.clusterService.getRoleType() === RoleType.FOLLOWER && this.isEnabled) {
            await this.ideConnector.applyState(state);
        }
    }

    public toggleSync() {
        this.isEnabled = !this.isEnabled;
        this.updateStatusBar();
        vscode.window.showInformationMessage(`One-IDE Sync: ${this.isEnabled ? 'Enabled' : 'Disabled'}`);
    }

    private updateStatusBar() {
        if (this.isEnabled) {
            this.statusBarItem.text = '$(sync) One-IDE: On';
            this.statusBarItem.tooltip = 'Click to disable synchronization';
        } else {
            this.statusBarItem.text = '$(sync-ignored) One-IDE: Off';
            this.statusBarItem.tooltip = 'Click to enable synchronization';
        }
    }

    public async start() {
        Logger.log(`Starting SyncService. SourceID: ${this.sourceId}`);
        // ClusterService starts automatically in constructor/init
    }

    public dispose() {
        this.isEnabled = false;
        this.unsubscribeRoleChange?.();
        this.unsubscribeLeaderUserActivity?.();
        this.statusBarItem.dispose();
    }
}
